io/zenoh-transport/src/unicast/establishment/ext/qos.rs

```diff
@@ -35,21 +41,226 @@ impl<'a> QoSFsm<'a> {
     }
 }
 
-/*************************************/
-/*              OPEN                 */
-/*************************************/
-#[derive(Clone, Copy, Debug, PartialEq, Eq)]
-pub(crate) struct StateOpen {
-    is_qos: bool,
+#[derive(Clone, Debug, PartialEq, Eq)]
+enum State {
+    NoQoS,
+    QoS {
+        reliability: Option<Reliability>,
+        priorities: Option<PriorityRange>,
+    },
+}
+
+impl State {
+    fn new(is_qos: bool, endpoint: &EndPoint) -> ZResult<Self> {
+        if !is_qos {
+            Ok(State::NoQoS)
+        } else {
+            let metadata = endpoint.metadata();
+
+            let reliability = metadata
+                .get(Metadata::RELIABILITY)
+                .map(Reliability::from_str)
+                .transpose()?;
+
+            let priorities = metadata
+                .get(Metadata::PRIORITIES)
+                .map(PriorityRange::from_str)
+                .transpose()?;
+
+            Ok(State::QoS {
+                priorities,
+                reliability,
+            })
+        }
+    }
+
+    fn try_from_u64(value: u64) -> ZResult<Self> {
+        match value {
+            0b000_u64 => Ok(State::NoQoS),
+            0b001_u64 => Ok(State::QoS {
+                priorities: None,
+                reliability: None,
+            }),
+            value if value & 0b110_u64 != 0 => {
+                let tag = value & 0b111_u64;
+
+                let priorities = if tag & 0b010_u64 != 0 {
+                    let start = Priority::try_from(((value >> 3) & 0xff) as u8)?;
+                    let end = Priority::try_from(((value >> (3 + 8)) & 0xff) as u8)?;
+
+                    Some(PriorityRange::new(start..=end))
+                } else {
+                    None
+                };
+
+                let reliability = if tag & 0b100_u64 != 0 {
+                    let bit = ((value >> (3 + 8 + 8)) & 0x1) as u8 == 1;
+
+                    Some(Reliability::from(bit))
+                } else {
+                    None
+                };
+
+                Ok(State::QoS {
+                    priorities,
+                    reliability,
+                })
+            }
+            _ => Err(zerror!("invalid QoS").into()),
+        }
+    }
+
+    /// Encodes [`QoS`] as a [`u64`].
+    ///
+    /// This function is used for encoding both of [`StateAccept`] in
+    /// [`crate::unicast::establishment::cookie::Cookie::ext_qos`] and
+    /// [`zenoh_protocol::transport::init::ext::QoS`].
+    ///
+    /// The three least significant bits are used to discrimnate five states:
+    ///
+    /// 1. QoS is disabled
+    /// 2. QoS is enabled but no priority range and no reliability setting are available
+    /// 3. QoS is enabled and priority range is available but reliability is unavailable
+    /// 4. QoS is enabled and reliability is available but priority range is unavailable
+    /// 5. QoS is enabled and both priority range and reliability are available
+    fn to_u64(&self) -> u64 {
```

Comment:
According to @Mallets only `StateAccept` could have a codec implementation as it's serialized on the wire. Otherwise this function is used for creating the `ZExtZ64` and `ZExtUnit` extensions. 

Other `establishment` extensions also just put their codec implementations in the same file. Even if everything were to have a codec I would still keep the `try_from_64`/`to_u64` functions because it's the same logic.